### ISSUE TITLE
Integrate into existing Jupyter-based workflows

### ISSUE THREAD

**Issue by bollwyvl** [jtpio/jupyterlite#41]
Awesome stuff here! Maybe these have been brought up elsewhere, if so, feel free to close!

## elevator pitch

Having an installable, python-based package with hooks for various other jupyter tools with a couple features would make it easy to adopt for workflows folk are already invested in.

## Use Cases

- as a Jupyter user, locally be able to interactively build up a custom jupyterlite site with contents, custom extensions, etc. and do whatever you need to with the files
- as an extension developer community, build a demo site in CI and deploy to somewhere (e.g. github pages)

### nbconvert

One of the jyve features I loved building is the [nbconvert exporter](https://github.com/deathbeds/jyve/blob/master/jyve/exporter.py). That code's pretty rough, but it _does the thing_ of marshalling all the static assets into a folder, ready to be committed to `gh-pages` (or dragged into SharePoint :scream:) for demos.

I'm imagining two, not necessarily mutually exclusive outputs of:
- a full editable lab
- a single Document that _looks_ like nbconvert, etc. but is built out of lab renderers
- a couple Documents, one per page

With those pieces, one could integrate nicely into other workflows like sphinx, jupyter-book, mkdocs, etc.

#### design ideas

- a python package 
  - an `entry_point` for `jupyterlite site`
  - this would make it show up in various CLI and API endpoints
  - all/some of the `jupyterlite` static assets as either...
    - `data_files` for "simplicity"
    - `package_data` it's gonna be _big_, duplication in wheels is no fun

As `nbconvert` likes to make _one_ file, the default output would probably most usefully be a `.zip` file... not so much fun for previewing, but quite handy for moving around (as again, it's gonna be big). But if it gets more complicated...

### labextension

There are bells and whistles that would need more tweaking than what the exposed nbconvert API offers. It would be lovely to have an in-juptyerlab(-classic) extension that built jupyterlite sites and then did... stuff with them.

#### design ideas

This part would _have_ to be in `data_files`, but wouldn't be very big.

- offer it as a new _File > export as... > custom jupyterlite_
  - open a new mainareawidget (or whatever makes sense in classic)
    - capture the advanced options as a JSON schema
      - make a JSON schema from ui
    - a drop area for adding additional contents paths

As to targets, I'm imagining other outputs/targets, such as:

- git bundle
- start hosting the site
  - just `/files`
  - serve on a federated content network, e.g. dat/ipfs
    - these would work quite well if the core assets (and labextensions folk might include) end up being very static, as they would get deduplicated in the network and locally

**Comment by bollwyvl** [comment]
Updated the description with a drawio mockup of the "kitchen sink" UI, though I'd imagine a lot of these could be hidden under an "advanced options" UI control. Really, you want to just drag contents, press button, get :candy: !

**Comment by bollwyvl** [comment]
So some progress here, nothing to share yet, but some notes:

## whither app?

to avoid a perilous build step of copying lots of things, i've at least temporarily moved the `app` folder into the python module:

```
setup.py
setup.cfg
src/
    jupyterlite/
        app/
```

with some minor changes to various `json` files, everything still seems to build correctly. However, I wonder if instead having the app webpack deploy itself into static makes more sense with `CopyPlugin` or whatever.

## entropy management

Some things I like in core, from a build perspective:
- exactly one `tsc -b` gets invoked against the private `@jupyterlab/meta`
  - a successful build yields a single `.tsbuildinfo`
- exactly one `webpack` gets run
  - the resulting `package.json` includes enough entropy to _know_ when the build has changed

It's probably worth getting some of these system properties back, and potentially exploiting them with a hash-based build-as-needed system like `doit` (though I'm sure there's something like it for node, if that's your poison).

**Comment by jtpio** [comment]
> However, I wonder if instead having the app webpack deploy itself into static makes more sense with `CopyPlugin` or whatever.

Yes that would be better. For now `app` builds itself just for sake of simplicity, but it would be nice to stage static assets separately.

> Some things I like in core, from a build perspective

This :+1: And others like watch mode.

In general it would be really cool to be able to reuse as much as possible from upstream JupyterLab when it comes to the build system. `@jupyterlab/buildutils` are already published and usable, so that's great. Thinking of scripts like integrity, releasing or updating dependencies.

**Comment by jtpio** [comment]
And thanks for sharing ideas and the suggestions :+1: 

It would indeed be great to integrate it with the rest of the Jupyter workflows. Distributing static assets via Python packages is also what the prebuilt extension does so it would make sense to do something similar.

But ideally this would be optional, so we can keep the deploy story as simple as possible (to just dumping files on a server for those who want to do just that). For example for now deploying on Vercel is just a matter of running `yarn && yarn run build` and setting `app` as the deployed directory.

Also having a workflow without Python can be more welcoming to other folks coming from a more traditional web dev background. Again this would not be an issue if optional.

**Comment by bollwyvl** [comment]
> And thanks for sharing ideas and the suggestions

well, you know the saying about opinions... but i'm sure we can agree PRs are better, even for talking about stuff... I'm just having to circle the beast before going for the kill. This issue already sounds like a couple PRs!

>  folks coming from a more traditional web dev background

Yes, that is an important group: those that _don't_ know jupyter but _do_ know web junk should be able to use this in their workflows once #7 hits. Indeed, the distinction between "app" and "packages" is probably an illusion, as it probably _all_ has to go up.

Those of us that _do_ know jupyter are the sods stuck _building_ all this filagreed junk on top of the specs and protocols, and #7 will help us, too.

But the audience _this_ issue is pointed at is the existing members of the jupyter community that have been crying out for an easy-to-share mechanism _of their existing interactive computing experience_ that doesn't involve npm, CI, kubernetes, etc. for _years_. True static, kernel-based publishing would actually mean you could deploy an interactive journal paper, etc. without the "productionizing" step.

If, as an analyst working in python on windows, an in-UI way to get a viable static bundle of _my work_ and the lab extensions i want but _without_ downloading 435mb of `node_modules`, that i can drag-and-drop into SharePoint... that's a big step forward.

**Comment by bollwyvl** [comment]
yeah. Some thoughts on a few packages:

## `jupyterlite`

| `pyproject.toml#/tool/flit/metadata/requires` | `pyproject.toml#/tool/flit/metadata/requires-extra` |
|-----------------------------------------------|-----------------------------------------------------|
| `jupyter_core`<br/>`jsonschema`<br/>`wheel`   | `nbconvert`<br/>`ipywidgets`<br/>`watchgod`         |

- build with `flit`
- depend on `jupyter_core` but basically nothing else from the stack
  - composite ordered config/data paths **hard requirement**
  - cargo cult the `.d` thing, again, _again_ from `notebook/jupyter_server`
    - or lobby for getting this into `jupyter_core` or `traitlets` :duck:
- provides
  - the ~20mb of required production static assets
    - _don't_ use `data_files` because duplication
    - > future: a `jupyterlite-dev-assets` which has source maps 
  - the CLI tools (as `JupyterApp`)
    - schema for a build file
    - `jupyter lite init (--json|--notebook) --interactive`
      - create a baseline config
      - optionally offer a dumb, step-by-step CLI GUI
        - more like `yo` than `cookiecutter`
    - `jupyter lite build`
      - read a description of a build (stored in a `.(json|ipynb)` 
      - build a working site
    - `jupyter lite validate`
      - look at a built site, make sure it's all self-consistent
    - `jupyter lite serve (--watch)`
      - dumb `http.server` with (optional) dumb polling
      - _never to used_ as a production server
    - `jupyter lite deploy`
      - dumb deployment options

  - python API of same
    - generate base types from schema
    - optionally provide widgets
  - we can _fudge_ the `(extra_)labextensions_path` with traitlets-by-name as a fallback
  - lazy `entry_points` for nbconvert for working with `.ipynb` metadata
    ```python
    "nbconvert.exporters": [
        "lite = jupyterlite.exporters:LiteExporter",            # lab and classic
        "lite-lab = jupyterlite.exporters:LabLiteExporter",     # no classic
        "lite-classic = jupyterlite.exporters:LabLiteExporter", # no lab
     ],
    ```


## `jupyterlite-editor`

| `setup.cfg#/options/install_requires`     | `package.json#/dependencies`                                                  |
|-------------------------------------------|-------------------------------------------------------------------------------|
| `jupyterlite`<br/>`jupyterlab`            | `@jupyterlab/application`<br/>`@deathbeds/jupyterlab-rjsf`<br/>`libarchivejs` |

- build with plain-old `setuptools`
  - ships the labextension in `data_files`
- provides
  - backend
    - nothing too novel, delegates to `nbconvert`, etc.
  - frontend
    - a full manifest of the hard-pinned `jupyterlite` assets
    - notebook/document editor to configure a site build description based on schema
    - live preview (`/files`)
    - runs _in_ JupyterLite 
      - docs site will be able to build a zipball of an interactively-configured lite site
      - accepts uploaded `.whl`
      - will fall down once the archive is over XXXmb, probably
        - good problem to have :laughing:

**Comment by bollwyvl** [comment]
(sort  of) have nbconvert working: https://github.com/jupyterlite/jupyterlite/pull/756#issuecomment-1214554290
